File: client/calc_client.c
// SPDX-License-Identifier: GPL-2.0
/*
 * calc_client (C) - terminal client for calc_server.
 *
 * Connects to the server's Unix domain socket, presents the same numbered
 * menu the Python client uses, and round-trips one calc_request /
 * calc_response per choice. Same wire format on the socket as the Python
 * server and the kernel chardev (24-byte request, 16-byte response), so
 * this C client interoperates with either Python or C server.
 *
 * "Request OKAY..." is a local-success signal printed after send() returns
 * without error - the wire protocol has no separate ACK message.
 */

#define _POSIX_C_SOURCE 200809L

#include <errno.h>
#include <getopt.h>
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>

#include "calc_proto.h"

#define DEFAULT_SOCKET "/tmp/calc_server.sock"

/* Menu entries. Order is independent of the protocol op codes - change
 * the table freely without touching the wire format. */
struct menu_entry {
    const char *label;
    int op;
};

static const struct menu_entry MENU[] = {
    {"Add 2 numbers",      CALC_OP_ADD},
    {"Subtract 2 numbers", CALC_OP_SUB},
    {"Multiply 2 numbers", CALC_OP_MUL},
    {"Divide 2 numbers",   CALC_OP_DIV},
};
#define MENU_SIZE (sizeof(MENU) / sizeof(MENU[0]))

/*
 * status_text() - map a calc_status code to a human-readable string.
 *
 * Mirrors the Python client's _STATUS_TEXT dict so the two clients
 * print the same friendly error wording.
 *
 * Inputs:
 *   @status: a CALC_STATUS_* value (typically from a calc_response).
 *
 * Returns:
 *   A static, NUL-terminated string. Never NULL.
 */
static const char *status_text(int32_t status)
{
    switch (status) {
    case CALC_STATUS_OK:       return "ok";
    case CALC_STATUS_BAD_OP:   return "unknown operation";
    case CALC_STATUS_DIV_ZERO: return "division by zero";
    default:                   return "error";
    }
}

/*
 * recv_exact() - read exactly @n bytes from a socket.
 *
 * Loops over recv() so a short read doesn't truncate the result.
 * EINTR is retried transparently.
 *
 * Inputs:
 *   @fd:  open socket.
 *   @buf: destination buffer; must hold at least @n bytes.
 *   @n:   number of bytes to read.
 *
 * Returns:
 *   1 on full read;
 *   0 if the peer closed the connection before @n bytes arrived;
 *   -1 on recv() error (errno is set).
 */
static int recv_exact(int fd, void *buf, size_t n)
{
    char *p = buf;
    while (n > 0) {
        ssize_t r = recv(fd, p, n, 0);
        if (r == 0) return 0;
        if (r < 0) {
            if (errno == EINTR) continue;
            return -1;
        }
        p += r;
        n -= (size_t)r;
    }
    return 1;
}

/*
 * read_line() - read one line of input from stdin into @buf.
 *
 * Strips the trailing newline. If the line is longer than the buffer,
 * drains the rest of the line (up to and including the next '\n') so
 * the next prompt doesn't inherit half a line of garbage.
 *
 * Inputs:
 *   @buf: destination buffer.
 *   @cap: capacity of @buf in bytes; must be > 1.
 *
 * Returns:
 *   1 on success (buf is NUL-terminated, newline stripped);
 *   0 on EOF before any data;
 *  -1 if the line exceeded @cap (excess characters were drained, the
 *     truncated prefix is left in @buf and the caller should reject it).
 */
static int read_line(char *buf, size_t cap)
{
    if (!fgets(buf, (int)cap, stdin)) return 0;
    size_t len = strlen(buf);
    if (len == 0) return 0;
    if (buf[len - 1] == '\n') {
        buf[len - 1] = '\0';
    } else if (len == cap - 1) {
        /* No newline AND we filled the buffer - drain the rest of the
         * line so the next prompt isn't pre-filled with garbage. */
        int ch;
        while ((ch = getchar()) != '\n' && ch != EOF) { }
        return -1;
    }
    return 1;
}

/*
 * parse_int64() - parse a signed 64-bit integer from a NUL-terminated string.
 *
 * Accepts decimal, 0x hex, 0b binary, 0o octal (via strtoll base 0).
 * Rejects strings that have any trailing non-numeric characters, so
 * "42abc" surfaces as an error rather than as 42.
 *
 * Inputs:
 *   @s:   source string (must be NUL-terminated).
 *   @out: where to write the parsed value on success.
 *
 * Returns:
 *   0 on success;
 *   -1 on parse failure or range overflow (errno is set on overflow).
 */
static int parse_int64(const char *s, int64_t *out)
{
    if (*s == '\0') return -1;
    char *end;
    errno = 0;
    long long v = strtoll(s, &end, 0);   /* base 0 -> accepts 0x.., 0b.., 0o.. */
    if (errno != 0 || end == s || *end != '\0') return -1;
    *out = (int64_t)v;
    return 0;
}

/*
 * do_request() - send one calc_request, print the calc_response.
 *
 * Mirrors the Python client's UX exactly: prints "Sending request..."
 * before send(), "Request OKAY..." after send() returns, then
 * "Receiving response..." before recv(), and finally either
 * "Result is N!" or the friendly status text for an error response.
 *
 * "Request OKAY..." is a *local* success signal — the wire protocol
 * carries no separate ACK message.
 *
 * Inputs:
 *   @sock: connected socket to the server.
 *   @op:   one of CALC_OP_* from calc_proto.h.
 *   @a, @b: signed 64-bit operands.
 *
 * Returns:
 *   0 on success (response received and printed, OK or error);
 *   -1 on transport error — the caller should bail out of the REPL.
 */
static int do_request(int sock, int op, int64_t a, int64_t b)
{
    struct calc_request req;
    memset(&req, 0, sizeof(req));        /* zeros _pad */
    req.op = op;
    req.a  = a;
    req.b  = b;

    printf("Sending request...\n");
    ssize_t sent = send(sock, &req, sizeof(req), MSG_NOSIGNAL);
    if (sent != (ssize_t)sizeof(req)) {
        fprintf(stderr, "  send failed: %s\n", strerror(errno));
        return -1;
    }
    printf("Request OKAY...\n");

    printf("Receiving response...\n");
    struct calc_response resp;
    int rc = recv_exact(sock, &resp, sizeof(resp));
    if (rc <= 0) {
        fprintf(stderr, "  connection closed mid-response\n");
        return -1;
    }

    if (resp.status == CALC_STATUS_OK) {
        printf("Result is %" PRId64 "!\n", (int64_t)resp.result);
    } else {
        printf("  %s\n", status_text(resp.status));
    }
    return 0;
}

/*
 * print_menu() - render the operation menu to stdout.
 *
 * Items are numbered 1..MENU_SIZE; the last slot (MENU_SIZE + 1) is
 * reserved for Exit.
 *
 * Inputs: (none)
 *
 * Returns: (void)
 */
static void print_menu(void)
{
    printf("\n");
    for (size_t i = 0; i < MENU_SIZE; i++) {
        printf("(%zu) %s\n", i + 1, MENU[i].label);
    }
    printf("(%zu) Exit\n", MENU_SIZE + 1);
}

/*
 * usage() - print short help to stderr.
 *
 * Inputs:
 *   @prog: argv[0], shown verbatim so the message reflects the user's
 *          invocation name.
 *
 * Returns: (void)
 */
static void usage(const char *prog)
{
    fprintf(stderr,
        "Usage: %s [--socket PATH]\n"
        "  --socket PATH   server socket (default: %s)\n",
        prog, DEFAULT_SOCKET);
}

/*
 * main() - parse args, connect to server, run the REPL loop.
 *
 * Connects to the server's Unix domain socket (default
 * /tmp/calc_server.sock); each iteration prints the menu, reads the
 * user's command and two operands, and calls do_request() to perform
 * one round-trip. Exits cleanly on option (MENU_SIZE + 1), on Ctrl-D,
 * or on transport error.
 *
 * Inputs:
 *   @argc, @argv: standard CLI args; --socket overrides the default
 *                 socket path.
 *
 * Returns:
 *   0 on clean exit;
 *   2 on usage error or unreachable server.
 */
int main(int argc, char *argv[])
{
    const char *socket_path = DEFAULT_SOCKET;

    static struct option opts[] = {
        {"socket", required_argument, NULL, 's'},
        {"help",   no_argument,       NULL, 'h'},
        {NULL, 0, NULL, 0},
    };
    int c;
    while ((c = getopt_long(argc, argv, "s:h", opts, NULL)) != -1) {
        switch (c) {
        case 's': socket_path = optarg; break;
        case 'h': usage(argv[0]); return 0;
        default:  usage(argv[0]); return 2;
        }
    }

    int sock = socket(AF_UNIX, SOCK_STREAM, 0);
    if (sock < 0) {
        perror("socket");
        return 2;
    }

    struct sockaddr_un addr;
    memset(&addr, 0, sizeof(addr));
    addr.sun_family = AF_UNIX;
    if (strlen(socket_path) >= sizeof(addr.sun_path)) {
        fprintf(stderr, "calc_client: socket path too long\n");
        close(sock);
        return 2;
    }
    strncpy(addr.sun_path, socket_path, sizeof(addr.sun_path) - 1);

    if (connect(sock, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
        fprintf(stderr,
            "calc_client: cannot reach %s: %s\n"
            "  Is the server running? Try ./scripts/start_py_server.sh\n",
            socket_path, strerror(errno));
        close(sock);
        return 2;
    }

    printf("Connected to %s\n", socket_path);

    char line[64];
    int rc = 0;
    for (;;) {
        print_menu();
        printf("Enter command: ");
        fflush(stdout);

        int r = read_line(line, sizeof(line));
        if (r == 0) { printf("\n"); break; }      /* EOF -> exit cleanly */
        if (r < 0)  { printf("  input too long\n"); continue; }
        if (line[0] == '\0') continue;

        char *end;
        long choice = strtol(line, &end, 10);
        if (*end != '\0') { printf("  not a number\n"); continue; }
        if (choice == (long)(MENU_SIZE + 1)) break;
        if (choice < 1 || choice > (long)MENU_SIZE) {
            printf("  pick 1..%zu\n", MENU_SIZE + 1);
            continue;
        }
        int op = MENU[choice - 1].op;

        printf("Enter operand 1: ");
        fflush(stdout);
        if (read_line(line, sizeof(line)) <= 0) continue;
        if (line[0] == '\0') continue;
        int64_t a;
        if (parse_int64(line, &a) < 0) {
            printf("  '%s' is not an integer\n", line);
            continue;
        }

        printf("Enter operand 2: ");
        fflush(stdout);
        if (read_line(line, sizeof(line)) <= 0) continue;
        if (line[0] == '\0') continue;
        int64_t b;
        if (parse_int64(line, &b) < 0) {
            printf("  '%s' is not an integer\n", line);
            continue;
        }

        if (do_request(sock, op, a, b) < 0) {
            rc = 2;
            break;
        }
    }

    close(sock);
    return rc;
}

File: client/calc_client.py
#!/usr/bin/env python3
"""calc_client - terminal client for calc_server.

Connects to the calc_server Unix domain socket, presents a numbered menu
of operations, and round-trips one calc_request/calc_response per choice.

Wire format: same 24-byte request / 16-byte response that calc_server
speaks (and that the kernel chardev speaks). The client does no
translation; it just packs the user's choices into a struct and unpacks
the response.
"""

from __future__ import annotations

import argparse
import socket
import struct
import sys

# Layouts must match server/module/calc_proto.h byte-for-byte.
REQUEST  = struct.Struct("=iiqq")    # op, _pad, a, b           -> 24 bytes
RESPONSE = struct.Struct("=iiq")     # status, _pad, result     -> 16 bytes

DEFAULT_SOCKET = "/tmp/calc_server.sock"

# Mirrors enum calc_op in calc_proto.h.
ADD, SUB, MUL, DIV = 1, 2, 3, 4

# Menu entries: (display label, op code). Order is independent of the op
# numbering — change it freely without touching the protocol.
MENU = [
    ("Add 2 numbers",      ADD),
    ("Subtract 2 numbers", SUB),
    ("Multiply 2 numbers", MUL),
    ("Divide 2 numbers",   DIV),
]

# Mirrors enum calc_status in calc_proto.h.
STATUS_OK       = 0
STATUS_BAD_OP   = 1
STATUS_DIV_ZERO = 2

_STATUS_TEXT = {
    STATUS_BAD_OP:   "unknown operation",
    STATUS_DIV_ZERO: "division by zero",
}


# --------------------------------------------------------------------------
# Wire helpers
# --------------------------------------------------------------------------

def _recv_exact(sock: socket.socket, n: int) -> bytes:
    """Read exactly n bytes from sock or raise ConnectionError."""
    buf = bytearray()
    while len(buf) < n:
        chunk = sock.recv(n - len(buf))
        if not chunk:
            raise ConnectionError("server closed connection mid-response")
        buf.extend(chunk)
    return bytes(buf)


def _do_request(sock: socket.socket, op: int, a: int, b: int) -> None:
    """One request/response round-trip with status-aware printing.

    Prints the same UX lines the spec's sample shows. "Request OKAY..."
    is a local-success signal (sendall() returned without error); the
    wire protocol has no separate ACK message and doesn't need one for
    this purpose.
    """
    print("Sending request...")
    sock.sendall(REQUEST.pack(op, 0, a, b))
    print("Request OKAY...")

    print("Receiving response...")
    status, _pad, result = RESPONSE.unpack(_recv_exact(sock, RESPONSE.size))

    if status == STATUS_OK:
        print(f"Result is {result}!")
    else:
        print(f"  {_STATUS_TEXT.get(status, f'error status {status}')}")


# --------------------------------------------------------------------------
# Terminal UI
# --------------------------------------------------------------------------

def _read_int(prompt: str) -> int | None:
    """Read a signed integer from stdin. None on EOF or invalid input.

    Accepts decimal, hex (0x...), octal (0o...), and binary (0b...) via
    int(x, 0). Range-checks against signed 64-bit so a too-large value
    surfaces as a friendly error rather than a struct.error from pack().
    """
    try:
        raw = input(prompt).strip()
    except EOFError:
        print()
        return None
    if not raw:
        return None
    try:
        value = int(raw, 0)
    except ValueError:
        print(f"  '{raw}' is not an integer")
        return None
    if not (-(1 << 63) <= value <= (1 << 63) - 1):
        print(f"  {value} is out of signed 64-bit range")
        return None
    return value


def _print_menu() -> None:
    print()
    for i, (label, _op) in enumerate(MENU, start=1):
        print(f"({i}) {label}")
    print(f"({len(MENU) + 1}) Exit")


def repl(sock: socket.socket) -> int:
    """Main interactive loop. Returns the process exit code."""
    while True:
        _print_menu()
        try:
            raw = input("Enter command: ").strip()
        except EOFError:
            print()
            return 0
        if not raw:
            continue
        try:
            choice = int(raw)
        except ValueError:
            print("  not a number")
            continue

        if choice == len(MENU) + 1:
            return 0
        if not 1 <= choice <= len(MENU):
            print(f"  pick 1..{len(MENU) + 1}")
            continue

        op = MENU[choice - 1][1]
        a = _read_int("Enter operand 1: ")
        if a is None:
            continue
        b = _read_int("Enter operand 2: ")
        if b is None:
            continue

        try:
            _do_request(sock, op, a, b)
        except (ConnectionError, OSError, struct.error) as exc:
            print(f"  connection error: {exc}")
            return 2


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="calc_dev terminal client")
    parser.add_argument(
        "--socket", default=DEFAULT_SOCKET,
        help=f"server socket path (default: {DEFAULT_SOCKET})",
    )
    args = parser.parse_args(argv)

    try:
        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.connect(args.socket)
    except (FileNotFoundError, ConnectionRefusedError) as exc:
        print(
            f"calc_client: cannot reach {args.socket}: {exc}\n"
            f"  Is the server running? Try ./scripts/start_py_server.sh",
            file=sys.stderr,
        )
        return 2

    print(f"Connected to {args.socket}")
    try:
        return repl(sock)
    except KeyboardInterrupt:
        print()
        return 0
    finally:
        sock.close()


if __name__ == "__main__":
    sys.exit(main())
